Common/Field.cs

```diff
@@ -24,28 +24,128 @@ namespace QuantConnect
     /// </summary>
     public static partial class Field
     {
+        private readonly static Func<IBaseData, decimal> _high = BaseDataBarPropertyOrValue(x => x.High);
+        private readonly static Func<IBaseData, decimal> _low = BaseDataBarPropertyOrValue(x => x.Low);
+        private readonly static Func<IBaseData, decimal> _open = BaseDataBarPropertyOrValue(x => x.Open);
+        private readonly static Func<IBaseData, decimal> _bidClose = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Bid.Close);
+        private readonly static Func<IBaseData, decimal> _bidOpen = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Bid.Open);
+        private readonly static Func<IBaseData, decimal> _bidLow = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Bid.Low);
+        private readonly static Func<IBaseData, decimal> _bidHigh = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Bid.High);
+        private readonly static Func<IBaseData, decimal> _askClose = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Ask.Close);
+        private readonly static Func<IBaseData, decimal> _askOpen = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Ask.Open);
+        private readonly static Func<IBaseData, decimal> _askLow = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Ask.Low);
+        private readonly static Func<IBaseData, decimal> _askHigh = BaseDataBarPropertyOrValue(x => ((QuoteBar)x).Ask.High);
+        private readonly static Func<IBaseData, decimal> _bidPrice = TickPropertyOrValue(x => ((Tick)x).BidPrice);
+        private readonly static Func<IBaseData, decimal> _askPrice = TickPropertyOrValue(x => ((Tick)x).AskPrice);
+        private readonly static Func<IBaseData, decimal> _volume = BaseDataBarPropertyOrValue(x => ((TradeBar)x).Volume);
+        private readonly static Func<IBaseData, decimal> _average = BaseDataBarPropertyOrValue(x => (x.Open + x.High + x.Low + x.Close) / 4m);
+        private readonly static Func<IBaseData, decimal> _median = BaseDataBarPropertyOrValue(x => (x.High + x.Low) / 2m);
+        private readonly static Func<IBaseData, decimal> _typical = BaseDataBarPropertyOrValue(x => (x.High + x.Low + x.Close) / 3m);
+        private readonly static Func<IBaseData, decimal> _weighted = BaseDataBarPropertyOrValue(x => (x.High + x.Low + 2 * x.Close) / 4m);
+        private readonly static Func<IBaseData, decimal> _sevenBar = BaseDataBarPropertyOrValue(x => (2 * x.Open + x.High + x.Low + 3 * x.Close) / 7m);
+
         /// <summary>
         /// Gets a selector that selects the Open value
         /// </summary>
         public static Func<IBaseData, decimal> Open
         {
-            get { return BaseDataBarPropertyOrValue(x => x.Open); }
+            get { return _open; }
```

Comment:
nit - leave High, Low and Close properties next to open just for clearness and organization